### PR TITLE
Document & test overlapping of key_exchange.

### PR DESCRIPTION
**Pull request by richwalm**
As discussed with Loup via e-mail, document and create a test for overlapping the shared_key & your_secret_key buffers for crypto_key_exchange().

Useful as in my application, there's no reason to keep around the secret key after obtaining a shared key, avoiding locking and wiping an extra buffer.

### CODE REVIEW

**Comment by fscoto** [comment]
The documentation change seems good to me. As for the test code, *maybe* it would be worthwhile to have a `buf[64]` and loop from 0 to 32, offsetting one of the arguments by `i` to test partial overlapping.

(Suggestion for in-tree change: Add the same wording to `crypto_x25519` as well.)

I'd like to request that you add yourself with a copyright line and to the "Written by" line for the files you've changed, or that you make an explicit comment here for @LoupVaillant to do that after merging, explicitly acknowledging that you subject your work to BSD-2-clause and CC-0.

File: tests/test.c [hunk 1]
```diff
@@ -660,6 +660,22 @@ static int p_argon2i_overlap()
...
+{
+    int status = 0;
+    FOR (i, 0, 5) {
+        u8 buf[32];
+        RANDOM_INPUT(sk, 32);
+        RANDOM_INPUT(pk, 32);
+        crypto_key_exchange(buf, sk, pk);
+        crypto_key_exchange(sk, sk, pk);
+        status |= memcmp(buf, sk, 32);
+    }
```
Comment (LoupVaillant) [inline tests/test.c]: I agree with @fscoto on this one, ideally we should have a bigger buffer and test every possible overlap (62 tests total?). While partial overlap don't make sense, making sure they are possible makes sure the code resist some overly clever uses.

So:

    FOR (i, 0, 64) {
        u8 overlapping[96];
        u8 separate[32];
        RANDOM_INPUT(sk, 32);
        RANDOM_INPUT(pk, 32);
        memcmp(overlapping+32, sk, 32);
        crypto_key_exchange(separate, sk, pk);
        crypto_key_exchange(overlapping + i, overlapping + 32, pk);
        status |= memcmp(separate, overlapping, 32);
    }

(Not tested, not checked the consistency with the rest of the test suite, but you get the idea.)

**Comment by richwalm** [comment]
Thanks. I've made those suggestions as mentioned.

**Comment by LoupVaillant** [comment]
Perfect, merging right away, thanks!

**Pull request merged**